File: core/stats_api/monthly_report.py
"""Monthly report data builder.

Generates statistics for the specified month:
- File count + month-over-month comparison
- Top 20 tags (with previous month rank comparison)
- New tags (first appeared in the current month)
- Source distribution
- Most active day + daily counts
- Trophy evaluation
- Available months list
"""

from core.extensions_core.service_registry import ServiceRegistry
from core.stats_api.filters import ai_image_where
from core.timezone_core.tz_helper import month_range_unix, tz_sqlite_modifier


def _month_range_unix(month_str: str) -> tuple[int, int]:
    """Convert 'YYYY-MM' to configured-timezone start/end Unix epoch."""
    return month_range_unix(month_str)


def _prev_month_str(month_str: str) -> str:
    """'2026-02' -> '2026-01'"""
    parts = month_str.split("-")
    year, mon = int(parts[0]), int(parts[1])
    if mon == 1:
        return f"{year - 1}-12"
    return f"{year}-{mon - 1:02d}"


def _top_tags_for_month(con, where: str, mstart: int, mend: int, limit: int = 20):
    """Retrieve top tags for the month. [(tag_id, tag, namespace, count), ...]

    Aggregates by tag_id (not tag name) to avoid merging same-name tags from
    different namespaces (e.g. general:hair vs artist:hair).
    Follows the 2-step pattern from basic.py: aggregate on file_tags first,
    resolve names after to avoid costly 3-way JOINs.
    """
    rows = con.execute(
        f"""SELECT ft.tag_id, COUNT(*) as cnt
            FROM file_tags ft
            JOIN files f ON f.id = ft.file_id
            WHERE {where} AND f.mtime >= ? AND f.mtime <= ?
            GROUP BY ft.tag_id
            ORDER BY cnt DESC
            LIMIT ?""",
        (mstart, mend, limit),
    ).fetchall()
    if not rows:
        return []
    ids = [r[0] for r in rows]
    placeholders = ",".join("?" * len(ids))
    name_map = {
        r[0]: (r[1], r[2])
        for r in con.execute(
            f"SELECT id, tag, namespace FROM tags WHERE id IN ({placeholders})", ids
        )
    }
    return [
        (tid, name_map[tid][0], name_map[tid][1], cnt)
        for tid, cnt in rows
        if tid in name_map
    ]


def _new_tags_for_month(con, where: str, mstart: int, mend: int, limit: int = 20) -> list[str]:
    """Retrieve tags that first appeared in the current month.

    Narrows candidates using the tags.first_seen_mtime index,
    then verifies actual usage via file_tags JOIN and returns in descending count order.
    """
    # Step 1: Get tag_id candidates first seen this month via first_seen_mtime (fast index lookup)
    tag_ids: list[int] = []
    tag_map: dict[int, str] = {}
    for tag_id, tag in con.execute(
        "SELECT id, tag FROM tags "
        "WHERE first_seen_mtime >= ? AND first_seen_mtime <= ?",
        (mstart, mend),
    ):
        tag_ids.append(tag_id)
        tag_map[tag_id] = tag
    if not tag_ids:
        return []

    # Step 2: Verify candidate tag_ids are actually linked to AI image files this month + count
    placeholders = ",".join("?" for _ in tag_ids)

    rows = con.execute(
        f"""SELECT ft.tag_id, COUNT(*) as cnt
            FROM file_tags ft
            JOIN files f ON f.id = ft.file_id
            WHERE ft.tag_id IN ({placeholders})
              AND {where}
              AND f.mtime >= ? AND f.mtime <= ?
            GROUP BY ft.tag_id
            ORDER BY cnt DESC
            LIMIT ?""",
        (*tag_ids, mstart, mend, limit),
    )
    return [tag_map[r[0]] for r in rows if r[0] in tag_map]


def _available_months(con, where: str) -> list[str]:
    """Return all months in the DB in descending order."""
    rows = con.execute(
        f"""SELECT DISTINCT strftime('%Y-%m',
                datetime(f.mtime, 'unixepoch', {tz_sqlite_modifier()!r})) as month
            FROM files f WHERE {where}
            ORDER BY month DESC"""
    )
    return [r[0] for r in rows if r[0]]


def build_monthly_report(con, month_str: str, *, include_trophies: bool = True) -> dict:
    """Build report data for the specified month."""
    where = ai_image_where("f")

    mstart, mend = _month_range_unix(month_str)
    prev_month = _prev_month_str(month_str)
    pstart, pend = _month_range_unix(prev_month)

    # --- File count (280K optimization: merge 2 COUNTs into 1 query) ---
    count_row = con.execute(
        f"""SELECT
            SUM(CASE WHEN f.mtime >= ? AND f.mtime <= ? THEN 1 ELSE 0 END),
            SUM(CASE WHEN f.mtime >= ? AND f.mtime <= ? THEN 1 ELSE 0 END)
        FROM files f WHERE {where}""",
        (mstart, mend, pstart, pend),
    ).fetchone()
    file_count = count_row[0] or 0
    prev_count = count_row[1] or 0

    # None when there is no previous month baseline (first month or gap after
    # a month with 0 files). The frontend renders None as "-- vs prev" via the
    # neutral branch, which is more accurate than "+0.0% vs prev".
    mom_pct: float | None = (
        round((file_count - prev_count) / prev_count * 100, 1) if prev_count > 0 else None
    )

    # --- Unique tag count (280K optimization: leverage cache) ---
    from core.stats_api.monthly_stats_materialize import get_cached_monthly_stat
    cached_tags = get_cached_monthly_stat(con, month_str, "unique_tags")
    if cached_tags is not None:
        unique_tags = int(cached_tags)
    else:
        unique_tags = con.execute(
            f"""SELECT COUNT(DISTINCT ft.tag_id)
                FROM file_tags ft
                JOIN files f ON f.id = ft.file_id
                WHERE {where} AND f.mtime >= ? AND f.mtime <= ?""",
            (mstart, mend),
        ).fetchone()[0]

    # --- Top tags + previous month rank ---
    # Both cur_tags and prev_tags are [(tag_id, tag, namespace, count), ...]
    # Keyed by tag_id to avoid merging same-name tags across namespaces.
    cur_tags = _top_tags_for_month(con, where, mstart, mend, 20)
    prev_tags = _top_tags_for_month(con, where, pstart, pend, 50)
    prev_rank_map = {row[0]: i + 1 for i, row in enumerate(prev_tags)}

    top_tags = []
    for rank, row in enumerate(cur_tags, 1):
        tid, tag, cnt = row[0], row[1], row[3]
        prev_rank = prev_rank_map.get(tid)
        rank_change = (prev_rank - rank) if prev_rank else None
        top_tags.append({
            "tag": tag, "count": cnt, "rank": rank,
            "prev_rank": prev_rank, "rank_change": rank_change,
        })

    # --- New tags (first appeared this month) ---
    new_tags = _new_tags_for_month(con, where, mstart, mend, 20)

    # --- Source distribution ---
    src_rows = con.execute(
        f"""SELECT COALESCE(f.meta_source, 'unknown') as src, COUNT(*) as cnt
            FROM files f WHERE {where} AND f.mtime >= ? AND f.mtime <= ?
            GROUP BY src ORDER BY cnt DESC""",
        (mstart, mend),
    )
    sources = {r[0]: r[1] for r in src_rows}

    # --- Most active day + daily counts ---
    daily_rows = con.execute(
        f"""SELECT date(f.mtime, 'unixepoch', {tz_sqlite_modifier()!r}) as day, COUNT(*) as cnt
            FROM files f WHERE {where} AND f.mtime >= ? AND f.mtime <= ?
            GROUP BY day ORDER BY day""",
        (mstart, mend),
    )
    daily_counts = [{"date": r[0], "count": r[1]} for r in daily_rows]

    most_active_day = None
    if daily_counts:
        best = max(daily_counts, key=lambda d: d["count"])
        most_active_day = {"date": best["date"], "count": best["count"]}

    # --- Trophies (ServiceRegistry fallback) ---
    trophies = []
    _judge_trophies = ServiceRegistry.get("trophies.judge_all") if include_trophies else None
    if _judge_trophies is not None:
        cumul_before = con.execute(
            f"SELECT COUNT(*) FROM files f WHERE {where} AND f.mtime < ?",
            (mstart,),
        ).fetchone()[0]
        cumul_after = cumul_before + file_count
        # Trophy evaluation requires writes (INSERT), so use a writable connection
        try:
            from core.services_core.db_api import get_db
            from core.services_core.db_write import submit_db_write

            def _write_trophies():
                write_con = get_db()
                trophies_local = _judge_trophies(
                    write_con, month_str, cumul_before, cumul_after,
                    daily_counts, sources, unique_tags_cumul=unique_tags,
                )
                if any(t.get("is_new") for t in trophies_local):
                    write_con.commit()
                return trophies_local

            trophies = submit_db_write(_write_trophies)
        except Exception:
            # If writable connection fails, skip with read-only fallback
            trophies = []

    # --- Available months ---
    months = _available_months(con, where)

    return {
        "month": month_str,
        "file_count": file_count,
        "prev_month_count": prev_count,
        "mom_change_pct": mom_pct,
        "unique_tags": unique_tags,
        "new_tags": new_tags,
        "top_tags": top_tags,
        "sources": sources,
        "most_active_day": most_active_day,
        "daily_counts": daily_counts,
        "trophies": trophies,
        "available_months": months,
    }

File: crates/yu-server/src/routes/monthly_report.rs
use axum::{
    extract::{Extension, Query, State},
    response::{IntoResponse, Response},
    Json,
};
use chrono::{Datelike, Local, NaiveDate, TimeZone};
use serde::Deserialize;
use serde_json::{json, Value};
use sqlx::{Row, SqlitePool};

use crate::{
    auth::{scope::require_admin_scope, AuthContext},
    state::SharedState,
};

const AI_WHERE: &str = "(
        f.is_deleted=0
        AND f.meta_source IS NOT NULL
        AND f.meta_source NOT IN ('', 'unknown', 'not_modified')
        AND f.meta_source NOT LIKE 'media_%'
    )";
const TZ_MODIFIER: &str = "localtime";

fn admin_scope_error(
    state: &SharedState,
    auth_context: Option<&Extension<AuthContext>>,
) -> Option<Response> {
    require_admin_scope(state.config.pin_auth_enabled, auth_context.map(|c| &c.0))
}

/// Returns (mstart_unix, mend_unix) inclusive bounds for YYYY-MM in local timezone.
fn month_range_unix(year: i32, month: u32) -> (i64, i64) {
    let last_day = last_day_of_month(year, month);
    let start = Local
        .with_ymd_and_hms(year, month, 1, 0, 0, 0)
        .single()
        .map(|dt| dt.timestamp())
        .unwrap_or(0);
    let end = Local
        .with_ymd_and_hms(year, month, last_day, 23, 59, 59)
        .single()
        .map(|dt| dt.timestamp())
        .unwrap_or(0);
    (start, end)
}

fn last_day_of_month(year: i32, month: u32) -> u32 {
    if month == 12 {
        NaiveDate::from_ymd_opt(year + 1, 1, 1)
    } else {
        NaiveDate::from_ymd_opt(year, month + 1, 1)
    }
    .and_then(|d| d.pred_opt())
    .map(|d| d.day())
    .unwrap_or(30)
}

fn prev_month(year: i32, month: u32) -> (i32, u32) {
    if month == 1 {
        (year - 1, 12)
    } else {
        (year, month - 1)
    }
}

async fn build_monthly_report(pool: &SqlitePool, month_str: &str) -> Result<Value, sqlx::Error> {
    let parts: Vec<&str> = month_str.split('-').collect();
    if parts.len() != 2 {
        return Ok(json!({"error": "invalid month"}));
    }
    let year: i32 = parts[0].parse().unwrap_or(0);
    let month: u32 = parts[1].parse().unwrap_or(0);
    if year == 0 || month == 0 || month > 12 {
        return Ok(json!({"error": "invalid month"}));
    }

    let (mstart, mend) = month_range_unix(year, month);
    let (py, pm) = prev_month(year, month);
    let (pstart, pend) = month_range_unix(py, pm);

    // File count (current + prev)
    let count_row = sqlx::query(&format!(
        r#"SELECT
            SUM(CASE WHEN f.mtime >= ? AND f.mtime <= ? THEN 1 ELSE 0 END),
            SUM(CASE WHEN f.mtime >= ? AND f.mtime <= ? THEN 1 ELSE 0 END)
           FROM files f WHERE {AI_WHERE}"#
    ))
    .bind(mstart)
    .bind(mend)
    .bind(pstart)
    .bind(pend)
    .fetch_one(pool)
    .await?;
    let file_count: i64 = count_row.try_get::<Option<i64>, _>(0)?.unwrap_or(0);
    let prev_count: i64 = count_row.try_get::<Option<i64>, _>(1)?.unwrap_or(0);
    let mom_pct: Option<f64> = if prev_count > 0 {
        Some(((file_count - prev_count) as f64 / prev_count as f64 * 100.0 * 10.0).round() / 10.0)
    } else {
        None
    };

    // Unique tags
    let unique_tags: i64 = sqlx::query(&format!(
        r#"SELECT COUNT(DISTINCT ft.tag_id)
           FROM file_tags ft JOIN files f ON f.id = ft.file_id
           WHERE {AI_WHERE} AND f.mtime >= ? AND f.mtime <= ?"#
    ))
    .bind(mstart)
    .bind(mend)
    .fetch_one(pool)
    .await?
    .try_get(0)?;

    // Top tags (2-step: aggregate then resolve names)
    let top_agg_rows = sqlx::query(&format!(
        r#"SELECT ft.tag_id, COUNT(*) as cnt
           FROM file_tags ft JOIN files f ON f.id = ft.file_id
           WHERE {AI_WHERE} AND f.mtime >= ? AND f.mtime <= ?
           GROUP BY ft.tag_id ORDER BY cnt DESC LIMIT 20"#
    ))
    .bind(mstart)
    .bind(mend)
    .fetch_all(pool)
    .await?;

    // Same for prev month top tags (for rank comparison)
    let prev_agg_rows = sqlx::query(&format!(
        r#"SELECT ft.tag_id, COUNT(*) as cnt
           FROM file_tags ft JOIN files f ON f.id = ft.file_id
           WHERE {AI_WHERE} AND f.mtime >= ? AND f.mtime <= ?
           GROUP BY ft.tag_id ORDER BY cnt DESC LIMIT 20"#
    ))
    .bind(pstart)
    .bind(pend)
    .fetch_all(pool)
    .await?;

    // Collect all tag_ids to resolve
    let cur_ids: Vec<i64> = top_agg_rows.iter().map(|r| r.get::<i64, _>(0)).collect();
    let prev_rank_map: std::collections::HashMap<i64, usize> = prev_agg_rows
        .iter()
        .enumerate()
        .map(|(i, r)| (r.get::<i64, _>(0), i + 1))
        .collect();

    let top_tags = if cur_ids.is_empty() {
        vec![]
    } else {
        let placeholders = cur_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
        let name_query = format!("SELECT id, tag FROM tags WHERE id IN ({placeholders})");
        let mut q = sqlx::query(&name_query);
        for id in &cur_ids {
            q = q.bind(id);
        }
        let name_rows = q.fetch_all(pool).await?;
        let name_map: std::collections::HashMap<i64, String> = name_rows
            .iter()
            .map(|r| (r.get::<i64, _>(0), r.get::<String, _>(1)))
            .collect();

        top_agg_rows
            .iter()
            .enumerate()
            .filter_map(|(i, r)| {
                let tid: i64 = r.get(0);
                let cnt: i64 = r.get(1);
                let tag = name_map.get(&tid)?.clone();
                let rank = i + 1;
                let prev_rank = prev_rank_map.get(&tid).copied();
                let rank_change = prev_rank.map(|p| p as i64 - rank as i64);
                Some(json!({
                    "tag": tag,
                    "count": cnt,
                    "rank": rank,
                    "prev_rank": prev_rank,
                    "rank_change": rank_change,
                }))
            })
            .collect::<Vec<_>>()
    };

    // New tags (first seen this month)
    let new_tag_candidates = sqlx::query(
        "SELECT id, tag FROM tags WHERE first_seen_mtime >= ? AND first_seen_mtime <= ?",
    )
    .bind(mstart)
    .bind(mend)
    .fetch_all(pool)
    .await?;

    let new_tags: Vec<String> = if new_tag_candidates.is_empty() {
        vec![]
    } else {
        let cand_ids: Vec<i64> = new_tag_candidates.iter().map(|r| r.get(0)).collect();
        let cand_map: std::collections::HashMap<i64, String> = new_tag_candidates
            .iter()
            .map(|r| (r.get::<i64, _>(0), r.get::<String, _>(1)))
            .collect();
        let placeholders = cand_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
        let verify_query = format!(
            r#"SELECT ft.tag_id, COUNT(*) as cnt
               FROM file_tags ft JOIN files f ON f.id = ft.file_id
               WHERE ft.tag_id IN ({placeholders})
                 AND {AI_WHERE} AND f.mtime >= ? AND f.mtime <= ?
               GROUP BY ft.tag_id ORDER BY cnt DESC LIMIT 20"#
        );
        let mut q = sqlx::query(&verify_query);
        for id in &cand_ids {
            q = q.bind(id);
        }
        q = q.bind(mstart).bind(mend);
        q.fetch_all(pool)
            .await?
            .iter()
            .filter_map(|r| cand_map.get(&r.get::<i64, _>(0)).cloned())
            .collect()
    };

    // Source distribution
    let src_rows = sqlx::query(&format!(
        r#"SELECT COALESCE(f.meta_source, 'unknown') as src, COUNT(*) as cnt
           FROM files f WHERE {AI_WHERE} AND f.mtime >= ? AND f.mtime <= ?
           GROUP BY src ORDER BY cnt DESC"#
    ))
    .bind(mstart)
    .bind(mend)
    .fetch_all(pool)
    .await?;
    let sources: serde_json::Map<String, Value> = src_rows
        .iter()
        .map(|r| (r.get::<String, _>(0), json!(r.get::<i64, _>(1))))
        .collect();

    // Daily counts
    let daily_rows = sqlx::query(&format!(
        r#"SELECT date(f.mtime, 'unixepoch', '{TZ_MODIFIER}') as day, COUNT(*) as cnt
           FROM files f WHERE {AI_WHERE} AND f.mtime >= ? AND f.mtime <= ?
           GROUP BY day ORDER BY day"#
    ))
    .bind(mstart)
    .bind(mend)
    .fetch_all(pool)
    .await?;
    let daily_counts: Vec<Value> = daily_rows
        .iter()
        .map(|r| {
            json!({
                "date": r.get::<Option<String>, _>(0).unwrap_or_default(),
                "count": r.get::<i64, _>(1),
            })
        })
        .collect();
    let most_active_day = daily_counts
        .iter()
        .max_by_key(|d| d["count"].as_i64().unwrap_or(0))
        .map(|d| {
            json!({
                "date": d["date"],
                "count": d["count"],
            })
        });

    // Available months
    let month_rows = sqlx::query(&format!(
        r#"SELECT DISTINCT strftime('%Y-%m', datetime(f.mtime, 'unixepoch', '{TZ_MODIFIER}')) as month
           FROM files f WHERE {AI_WHERE} ORDER BY month DESC"#
    ))
    .fetch_all(pool)
    .await?;
    let available_months: Vec<String> = month_rows
        .iter()
        .filter_map(|r| r.try_get::<Option<String>, _>(0).ok().flatten())
        .collect();

    Ok(json!({
        "month": month_str,
        "file_count": file_count,
        "prev_month_count": prev_count,
        "mom_change_pct": mom_pct,
        "unique_tags": unique_tags,
        "new_tags": new_tags,
        "top_tags": top_tags,
        "sources": sources,
        "most_active_day": most_active_day,
        "daily_counts": daily_counts,
        "trophies": [],
        "available_months": available_months,
    }))
}

#[derive(Deserialize)]
pub struct MonthlyReportParams {
    month: Option<String>,
}

pub async fn monthly_report(
    State(state): State<SharedState>,
    auth_context: Option<Extension<AuthContext>>,
    Query(params): Query<MonthlyReportParams>,
) -> Response {
    if let Some(err) = admin_scope_error(&state, auth_context.as_ref()) {
        return err;
    }

    let month = params.month.unwrap_or_else(|| {
        let now = Local::now();
        format!("{:04}-{:02}", now.year(), now.month())
    });

    // Validate YYYY-MM format
    let valid = month.len() == 7
        && month.chars().enumerate().all(|(i, c)| match i {
            0..=3 | 5..=6 => c.is_ascii_digit(),
            4 => c == '-',
            _ => false,
        });
    if !valid {
        return Json(json!({
            "ok": false,
            "error": "Invalid month format (expected YYYY-MM)",
            "data": null,
        }))
        .into_response();
    }

    match build_monthly_report(&state.db_read, &month).await {
        Ok(data) => Json(json!({"ok": true, "error": null, "data": data})).into_response(),
        Err(e) => {
            tracing::error!("monthly_report error: {e}");
            Json(json!({"ok": false, "error": "Database error", "data": null})).into_response()
        }
    }
}
